task_2/join.go

```diff
@@ -0,0 +1,124 @@
+package task_2
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+func Start() {
+
+	var (
+		numOfChannels = rand.Intn(12) + 1
+		channels      []chan interface{}
+	)
+
+	fmt.Println(fmt.Sprintf("creating %d channels..", numOfChannels))
+	for i := 0; i < numOfChannels; i++ {
+		channels = append(channels, make(chan interface{}))
+	}
+
+	ctx, _ := context.WithTimeout(context.Background(), 35*time.Second)
+
+	joinedChan := join(ctx, channels...)
+	go writeToChannels(channels)
+	for {
+		select {
+		case <-time.After(10 * time.Second):
+			fmt.Println("reached time limit waiting for joined output, exiting.")
+			return
+		case joinedOutput, ok := <-joinedChan:
+			if !ok {
+				return
+			}
+			fmt.Println("[RECEIVED JOINED OUTPUT]: ", joinedOutput)
+		}
+	}
+
+}
+
+func writeToChannels(ch []chan interface{}) {
+	for i := 1; i < 10; i++ {
+		fmt.Println(fmt.Sprintf("-- CYCLE %d -- sending messages to channels..", i))
+		for j, c := range ch {
+			c <- fmt.Sprintf("<cycle %d hi_%d>", i, j)
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func join(ctx context.Context, channels ...chan interface{}) chan interface{} {
+	joinedChan := make(chan interface{})
+	beginNewCycleChan := make(chan interface{}, 6)
+	output := make(chan interface{})
+	var wg sync.WaitGroup
+	wg.Add(len(channels))
+	messages := make([]interface{}, 0, len(channels))
+
+	for i, c := range channels {
+		fmt.Println("dispatching goroutine to receive from channel..")
+		go func(ctx context.Context, input <-chan interface{}, beginNewCycleChan chan interface{}, output chan interface{}, index int) {
+			defer wg.Done()
+			var sentMessageThisCycle bool
+			for {
+
+				if !sentMessageThisCycle {
+					select {
+					case <-ctx.Done():
+						fmt.Println("context Done() received, closing goroutine")
+						return
+					default:
+						val := <-input
+						output <- val
+						sentMessageThisCycle = true
+						fmt.Println("[GOROUTINE SENT OUTPUT]")
+					}
+
+				} else {
+					fmt.Println("[GOROUTINE WAITING] for new cycle..")
+					select {
+					case _, ok := <-beginNewCycleChan:
+						if ok {
+							sentMessageThisCycle = false
+						}
+					case <-ctx.Done():
+						fmt.Println("context Done() received, closing goroutine")
+						return
+					}
+				}
+
+			}
+		}(ctx, c, beginNewCycleChan, output, i)
+	}
+	go func() {
+		for {
+			select {
+			case <-time.After(60 * time.Second):
+				fmt.Println("timeout..")
+				return
+			case outputMessage, ok := <-output:
```

Comment:
Since we always send new values to the channels synchronically, our output is predictable and seems to be sequential in the order we send messages. It is better to check the correctness of your solution with randomized intervals of input data as well.